Add participants to on chain privacy group

iikirilov opened the pull request:
<!-- Thanks for sending a pull request! Please check out our contribution guidelines: -->
<!-- https://github.com/hyperledger/besu/blob/master/CONTRIBUTING.md -->

## PR description

## Fixed Issue(s)
<!-- Please link to fixed issue(s) here using format: fixes #<issue number> -->
<!-- Example: "fixes #2" -->

iikirilov committed:
onchain privacy group management

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
onchain privacy group management

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

josh-richardson committed:
Fixes build warnings & errors

Signed-off-by: Joshua Richardson <joshua@richardson.tech>

josh-richardson committed:
Fixes gas usage

Signed-off-by: Joshua Richardson <joshua@richardson.tech>

iikirilov committed:
add privxFindPrivacyGroup, refactor and remove privxCreatePrivacyGroup

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
fix bug

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
address comments

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
spotless

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
fix test

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
function to build list of transactions for privacy group

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
improve AT

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
Fix transaction receipt

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
Test contract can be deployed

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
more tests

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
refactor

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
add license

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
- retrieve private payload from pmt hash
- serialize add to group payload
- deserialize add to group payload

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
break

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:
```diff
@@ -0,0 +1,141 @@
+                .getPrivateTransactionMetadataList().stream()
+                .map(PrivateTransactionMetadata::getPrivacyMarkerTransactionHash)
+                .collect(Collectors.toList()));
+        blockHash = blockchain.getBlockHeader(blockHash).get().getParentHash();
```
You are working your way from the head down to the first PMT. I think you end up serialising the data starting with the last PMT/PTx. It might be easier for the new member and a bit more logical if the order of the serialised data was from first PMT to the last PMT ...

pinges commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java:
```diff
@@ -63,6 +76,29 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    final String addPayloadEnclaveKey;
+    if (privacyController.isGroupAdditionTransaction(privateTransaction)) {
+      final List<Hash> hashes =
+          onChainPrivacyController.buildTransactionList(
```
You have got these three methods in the OnChainPrivacyController, why don't you have another method there that soes all the work and just returns the "addPayloadEnclaveKey"?

pinges commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/priv/PrivGetTransactionReceipt.java:
```diff
@@ -106,7 +107,23 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+          privacyGroup =
+              privacyController.retrievePrivacyGroup(
+                  privateTransaction.getPrivacyGroupId().get().toBase64String());
+        } catch (final EnclaveClientException e) {
```
We should know whether the user wants to use the onchain functionality or not and we should use that information here to make that code clearer.

iikirilov committed:
fix AT

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
spotless

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
spotless

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
refactor blob to send to new participant

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
post-rebase fix

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
fix test

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
lock management before adding new participants

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
- check in precompile that contract is locked before adding participants
- check in precompile that contract is unlocked before running a transaction which is not adding participants

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
- disable Orion logs in AT

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
spotless

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

iikirilov committed:
add versioning

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

pinges commented on acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java:
```diff
@@ -24,21 +24,34 @@
+        node.execute(transactions.getPrivateTransactionReceipt(transactionHash));
+    assertThat(actualReceipt)
+        .usingRecursiveComparison()
+        .ignoringFields("commitmentHash", "logs")
```
why can the pmt hash not be compared here?

iikirilov replied:
because the commitment hash will be different every time as the pg id is random

pinges commented on acceptance-tests/dsl/src/main/java/org/hyperledger/besu/tests/acceptance/dsl/privacy/condition/ExpectValidPrivateTransactionReceipt.java:
```diff
@@ -24,21 +24,34 @@
+    for (int i = 0; i < expectedReceipt.getLogs().size(); i++) {
+      assertThat(actualReceipt.getLogs().get(i))
+          .usingRecursiveComparison()
+          .ignoringFields("blockHash", "blockNumber")
```
why ignore the blockHash and the blockNumber?

iikirilov replied:
because don't know which block it will be in

pinges commented on acceptance-tests/tests/src/test/java/org/hyperledger/besu/tests/web3j/privacy/OnChainPrivacyAcceptanceTest.java:
```diff
@@ -0,0 +1,353 @@
+        privateTransactionVerifier.validOnChainPrivacyGroupExists(
+            expectedGroupAfterCharlieIsAdded));
+
+    final Optional<TransactionReceipt> aliceAddReceipt =
```
I think to really test that charlie receives all the necessary transactions up to and including the add-transaction and all the transactions after the add we need to "flood" the transaction pool with private transactions before and after the "pause", and before and after the "add". Then we can go through all the receipts and check that all receipts for successful transactions are available to charlie and the same as the ones that alice and bob create.

iikirilov replied:
it is hard to do this in the AT

iikirilov committed:
break

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

pinges commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/privx/PrivxFindPrivacyGroup.java:
```diff
@@ -0,0 +1,66 @@
+
+  @Override
+  public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+    LOG.trace("Executing {}", RpcMethod.PRIVX_FIND_PRIVACY_GROUP.getMethodName());
```
With multi-tenancy we have to check whether the current user is allowed to find these privacy groups by checking whether the current user is one of the addresses ... But I think multi-tenancy is not merged in yet?

iikirilov replied:
multi-tenancy will introduce complications

iikirilov committed:
break

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

josh-richardson committed:
Fixes comparison key length bug, still has lots of nice bugs

iikirilov committed:
break

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java:
```diff
@@ -0,0 +1,536 @@
+          final PrivateTransactionProcessor.Result result =
+              privateTransactionProcessor.processTransaction(
+                  currentBlockchain,
+                  publicWorldState,
```
The public state to use here is the state at the end of the public Tx just before the PMT that the private Tx was marked by!

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/PrivacyPrecompiledContract.java:
```diff
@@ -129,7 +129,9 @@ public Bytes compute(final Bytes input, final MessageFrame messageFrame) {
     final Bytes32 privacyGroupId =
-        Bytes32.wrap(Bytes.fromBase64String(receiveResponse.getPrivacyGroupId()));
+        privateTransaction.getPrivacyGroupId().isPresent()
+            ? Bytes32.wrap(privateTransaction.getPrivacyGroupId().get())
```
I cannot see why this change is necessary ...

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:
```diff
@@ -0,0 +1,132 @@
+        privateStateStorage
+            .getPrivacyGroupHeadBlockMap(blockchain.getChainHeadHash())
+            .orElse(PrivacyGroupHeadBlockMap.EMPTY);
+    if (privacyGroupHeadBlockMap.get(privacyGroupId) != null) {
```
What happens when I'm trying to add someone to a privacy group that does not exist? What should happen? Should we fail?

iikirilov replied:
nope - it is fine - will create a new group

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:
```diff
@@ -0,0 +1,132 @@
+    PrivacyGroupHeadBlockMap privacyGroupHeadBlockMap =
+        privateStateStorage
+            .getPrivacyGroupHeadBlockMap(blockchain.getChainHeadHash())
+            .orElse(PrivacyGroupHeadBlockMap.EMPTY);
```
This means that I'm trying to add someone to a privacy group that does not exist! What is supposed to happen in that case?

iikirilov replied:
a new group is created

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/OnChainPrivacyController.java:
```diff
@@ -0,0 +1,132 @@
+        }
+      }
+    }
+    return pmtHashes;
```
variable name? It is not only PMT hashes now!

pinges commented on ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/privacy/methods/eea/EeaSendRawTransaction.java:
```diff
@@ -63,6 +77,31 @@ public JsonRpcResponse response(final JsonRpcRequestContext requestContext) {
+      final List<PrivateTransactionMetadata> privateTransactionMetadataList =
+          onChainPrivacyController.buildTransactionMetadataList(
+              Bytes32.wrap(privateTransaction.getPrivacyGroupId().get()));
+      if (privateTransactionMetadataList.size() > 0) {
```
maybe we should fail if there are no transactions ...

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:
```diff
@@ -92,8 +127,64 @@ public SendTransactionResponse sendTransaction(
+          privateStateStorage
+              .getPrivacyGroupHeadBlockMap(blockHash)
+              .orElse(PrivacyGroupHeadBlockMap.EMPTY);
+      final List<Bytes32> addDataKey =
```
should that happen in a PrivacyController? Should that not only happen in the OnChainPrivacyController?

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:
```diff
@@ -92,8 +127,64 @@ public SendTransactionResponse sendTransaction(
+              .getPrivacyGroupHeadBlockMap(blockHash)
+              .orElse(PrivacyGroupHeadBlockMap.EMPTY);
+      final List<Bytes32> addDataKey =
+          privacyGroupHeadBlockMap.keySet().stream()
```
... and why is that code different from the code in the OnChainPrivacyControler?

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:
```diff
@@ -113,11 +204,40 @@ public String deletePrivacyGroup(final String privacyGroupId, final String encla
     return enclave.findPrivacyGroup(addresses);
   }
 
+  public List<PrivacyGroup> findOnChainPrivacyGroup(
```
We have to make sure they only get groups back that they are part of. "They" is the enclavePublicKey here. We should check whether "they" are part of the addresses!

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:
```diff
@@ -164,25 +284,143 @@ public long determineNonce(
   private SendResponse sendRequest(
       final PrivateTransaction privateTransaction, final String enclavePublicKey) {
+
+    final Optional<PrivateTransactionProcessor.Result> process =
```
s/process/result/

pinges replied:
Could we put the code that is working out the version into a seperate method?

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/privacy/PrivacyController.java:
```diff
@@ -164,25 +284,143 @@ public long determineNonce(
+    process.ifPresent(
+        result -> {
+          if (!result.getOutput().toHexString().equals("0x")) {
+            rlpOutput.writeBytes(result.getOutput());
```
I'm not sure I like to have the "version" just appended to the end of the payload/private transaction. What are the other options that we have? Maybe just append it to the privacy group id? 
Or should we just encode the payload in a nicer way, so that it contains PGID, version, and the private transaction?
We do not have to add the version to the "legacy" private transactions, because the private transaction contains all the recipients, and the 'pantheon' style private transaction don't need it either, because the members cannot change.

iikirilov replied:
i don't feel that this should be a priority - I believe the impact of refactoring this later will be small

iikirilov committed:
fix public state bug

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>

pinges commented on ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/precompiles/privacy/OnChainPrivacyPrecompiledContract.java:
```diff
@@ -223,6 +224,40 @@ public Bytes compute(final Bytes input, final MessageFrame messageFrame) {
+
+          // There are the transaction that need to be run in the context of the world state with
+          // state root initialStateRoot
+          final List<Transaction> transactionsToRun =
```
If there were more than one PMT in the block you would run the same transactions (and more) for the next PMT ...

pinges committed:
push for Ivaylo

Signed-off-by: Stefan Pingel <stefan.pingel@consensys.net>

iikirilov committed:
break

Signed-off-by: Ivaylo Kirilov <iikirilov@gmail.com>
